rclcpp/include/rclcpp/create_timer.hpp

```diff
@@ -32,8 +32,8 @@ namespace rclcpp
 template<typename CallbackT>
 typename rclcpp::TimerBase::SharedPtr
 create_timer(
-  node_interfaces::NodeBaseInterface * node_base,
-  node_interfaces::NodeTimersInterface * node_timers,
+  std::shared_ptr<node_interfaces::NodeBaseInterface> node_base,
+  std::shared_ptr<node_interfaces::NodeTimersInterface> node_timers,
```

Comment:
I think this is reasonable, but we will also break a lot of downstream packages if the old function returning a raw pointer is deleted.
IMO, we should deprecate instead of delete it.

Reply:
I agree, and we could so for `rclcpp::create_timer`, but not for the generated API (at least not seamlessly, template argument deduction won't be carried out for return types). #1035, on the other hand, doesn't drop API. Ultimately, there's a discussion about ownership that has to settled on.

Reply:
> Ultimately, there's a discussion about ownership that has to settled on.

I think that it's up to the class requiring the interfaces to decide if they need shared ownership or just a non-owning pointer.

IMO, we should only provide getters returning shared pointers, and deprecate the old ones returning raw pointers (which can be replaced with `get_shared_*_interface().get()`).
But deleting that signature now will require a lot of changes in downstream packages.

> I agree, and we could so for rclcpp::create_timer, but not for the generated API (at least not seamlessly, template argument deduction won't be carried out for return types)

I don't understand why it can be done in the generated API, I will check again tomorrow.

Reply:
It might not look pretty, but we should be able to write some empy conditions to check if we're generating getters for one of the previously existing interfaces and insert the deprecated API.

Reply:
> I agree, and we could so for rclcpp::create_timer, but not for the generated API (at least not seamlessly, template argument deduction won't be carried out for return types).

Yes, template argument deduction won't work, but the same applies for the PR creating the files manually.
Both functions need to have a different name.

Reply:
Precisely, that's what #1035 does. We can make this PR do the same, though it'll take two deprecation cycles for the `std::shared_ptr` flavor API to get the name we intended it to have here.

Reply:
> We can make this PR do the same, though it'll take two deprecation cycles for the std::shared_ptr flavor API to get the name we intended it to have here.

That's a great point.
If you want to continue with the API breaking version, I can help with reviewing the PRs needed in the downstream packages.
It would also be good to add a comment in the release notes.

If you want to continue with two functions with different names, and optionally deprecate the ones returning a raw pointer, I also approve it.

Reply:
@jacobperron we need an executive decision here.

Reply:
I see. I'd rather not go through two deprecation cycles to get the API we want. I think it will be okay to have this API break and document it in the release notes.